File: Striver/Strings/20_Valid_Paranthesis.py
s = "([{}])"


def validPal(s):
    stack = []
    for i in range(len(s)):

        # empty condition needed for case with only closing brackets. You can't pop an empty list
        if not stack or s[i] == "(" or s[i] == "{" or s[i] == "[":
            stack.append(s[i])
        else:
            if s[i] == ")" and stack.pop() != "(":
                return False
            if s[i] == "}" and stack.pop() != "{":
                return False
            if s[i] == "]" and stack.pop() != "[":
                return False

    # If stack is not empty
    if stack:
        return False

    return True


print(validPal(s))
print(chr(65))

File: Striver/Strings/20_Valid_Paranthesis.cpp
#include <bits/stdc++.h>

using namespace std;

bool checkVal(string s)
{
  stack<int> st;
  for (int i = 0; i < s.size(); i++)
  {
    if (st.empty() || s[i] == '{' || s[i] == '(' || s[i] == '[')
    {
      st.push(s[i]);
    }
    else if (s[i] == '}')
    {
      if (st.top() != '{')
        return false;
      else
        st.pop();
    }
    else if (s[i] == ']')
    {
      if (st.top() != '[')
        return false;
      else
        st.pop();
    }
    else if (s[i] == ')')
    {
      if (st.top() != '(')
        return false;
      else
        st.pop();
    }
  }

  if (st.empty())
    return true;
  else
    return false;
}

int main()
{
  string s = "([{}])";
}